### PR TITLE
Simulator ng

### PR DESCRIPTION
**Pull request by slosar**
A new generation of simulation with jax and all the goodies.

### CODE REVIEW

**Comment by akvydula** [comment]
Updates as of 2/19: 

- `CroSimulator`  uses the same time and loc from `lusee.observation` class, same beam from `lusee.beam` class as `DefaultSimulator`
- beam rotated to alm and used a `for` loop for each beam
- note that Christian's example for croissant uses normalization factor of `total power` instead of `4pi` used in `DefaultSimulator`. `CroSimulator` uses `4pi`
- the config file now has `engine` where `Default` or `Croissant` can be specified. The most updates run script is `simulation/driver/run_Cro_sim.py`
- used Claude to write a test script `tests/test_cro_vs_default.py` that uses the last 5 time stamps and compares the output from each step of the simulation and shows where the two engines diverge. As of today, they diverge in which frame the computation is performed.
`- simulation/output/plot_default_vs_cro_sims.ipynb` just plots the results from both the outputs.

File: lusee/Beam.py [hunk 1]
```diff
@@ -486,7 +488,87 @@ def copy_beam(self,Etheta=None, Ephi=None):
...
+        ``(alt, az, freq)`` triplet.
+
+        The coordinate mapping is: ``theta = pi/2 - alt``, ``phi = az mod 2*pi``.
+
+        :returns: ``(interp_Etheta, interp_Ephi)``, each with signature
+            ``f(alt, az, freq) -> complex scalar or ndarray`` whose shape matches
+            the broadcast shape of the inputs.
+        :rtype: tuple(callable, callable)
+        """
+        from scipy.interpolate import RegularGridInterpolator
```
Comment (xzackli) [inline lusee/Beam.py]: should move this out to top level, we have a hard scipy dep

File: lusee/Beam.py
Comment (xzackli) [inline lusee/Beam.py:72]: wait what, why do we have this in here? this is a bespoke SHT?

File: lusee/BeamGauss.py [hunk 2]
```diff
@@ -1,11 +1,10 @@
 from .Beam import Beam
 import numpy as np
 
-def gauss_beam(theta,phi,sigma, theta_c,phi_c=0.):
+def gauss_beam(dec,phi,sigma, dec_c,phi_c=0.):
     """
     Function that creates a map-level gaussian beam in theta and phi of width sigma, centered at theta=theta_c and phi=phi_c.
```
Comment (xzackli) [inline lusee/BeamGauss.py:7]: Struggling to comprehend this. So this is "declination in beam coordinates", but i.e it's 90 degrees minus zenith angle? Default simulator lives in topo and then this is the complementary angle to the angle of the beam relative to zenith along the meridian?

For the purposes of a review, would be interested in either some firm documentation on this in the comments and docstring, or switch to a more conventional angle like just zenith angle along meridian.

File: lusee/CroSimulator.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import jax.numpy as jnp
+import croissant as cro
+import croissant.jax as crojax
+import jax
+from lunarsky import LunarTopo
+import s2fft
+
+"""
+CroSimulator: same inputs as DefaultSimulator (beam, sky, obs, etc.) but uses
+the Croissant engine for the actual simulation (MCMF frame, rot_alm_z phases,
```
Comment (xzackli) [inline lusee/CroSimulator.py]: We should find and replace all MCMF to MEPA!

File: lusee/CroSimulator.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+crojax.simulator.convolve). Freq, time range, and antenna location come from
+the observation object (config). Croissant currently supports single
+polarization / single dipole per beam; one beam combination at a time
+"""
+
+
+def healpy_packed_alm_to_croissant_2d(packed_alm, lmax):
```
Comment (xzackli) [inline lusee/CroSimulator.py]: I wonder if s2fft has this function somewhere, a good task for Claude

File: lusee/CroSimulator.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        delta_t = float(self.obs.deltaT_sec)
+
+        if self.sky_model.frame != "galactic":
+            raise NotImplementedError(
+                f"CroSimulator requires galactic sky frame, got {self.sky_model.frame}"
+            )
+        self.result = self._simulate_croissant_mepa(times, ntimes, delta_t)
+        return self.result
+
+    def _simulate_croissant_mepa(self, times, ntimes, delta_t):
```
Comment (xzackli) [inline lusee/CroSimulator.py:88]: will need to adjust this to be less object-like, but I think can happen in the next PR

File: setup_env.sh [hunk 6]
```diff
@@ -23,7 +23,8 @@ ljupyter_dev() { port=9600; docker run  -e HOME $PYTHON_SETUP $DEV_MOUNT -w $PWD
 
 # For your own environment, set these variables as shown in these examples
 
-# LUSEEPY_PATH="/home/user/work/lusee/luseepy"
-# REFSPEC_PATH="/home/user/work/lusee/luseepy"
-# LUSEE_DRIVE_DIR="/home/user/work/lusee/Drive/"
-# LUSEE_OUTPUT_DIR="/home/user/work/lusee/luseepy/simulation/output" 
+LUSEEPY_PATH="/Users/akshatha.vydula/lusee/luseepy"
```
Comment (xzackli) [inline setup_env.sh]: please don't commit these

**Review by xzackli — CHANGES_REQUESTED** [review]

File: lusee/CroSimulator.py (hunk 4 above)
Comment (akvydula) [reply lusee/CroSimulator.py]: yup! s2fft.sampling.reindex.flm_hp_to_2d_fast --updated in the latest commit

File: setup_env.sh (hunk 6 above)
Comment (akvydula) [reply setup_env.sh]: added a generic setup.sh file and a local untracked version of it. setup also nw can override the generic if the user creates a local file

File: lusee/Beam.py
Comment (akvydula) [reply lusee/Beam.py:72]: I am not sure why this is here, but my guess is that this function only gets used once so perhaps never got correctly placed?

File: lusee/BeamGauss.py (hunk 2 above)
Comment (akvydula) [reply lusee/BeamGauss.py:7]: add a few lines to doctoring, but @slosar can confirm the details

**Review by xzackli — APPROVED** [review]

File: lusee/BeamGauss.py (hunk 2 above)
Comment (slosar) [reply lusee/BeamGauss.py:7]: Ok, I fixed it. BeamGauss uses alt/az convention, gauss_beam helper function uses theta/phi since this is what Beam object uses internally.
